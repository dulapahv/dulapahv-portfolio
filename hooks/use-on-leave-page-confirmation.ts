'use client';

import { useEffect, useRef } from 'react';
import { usePathname, useRouter } from 'next/navigation';

const MESSAGE = 'Changes you made may not be saved.';

export function useOnLeavePageConfirmation(shouldConfirm: boolean) {
  const pathname = usePathname();
  const router = useRouter();

  const isInitialMount = useRef(true);

  // Handle refresh or close tab/window
  useEffect(() => {
    if (!shouldConfirm) return;

    function beforeUnload(e: BeforeUnloadEvent) {
      e.preventDefault();
      e.returnValue = '';
    }

    window.addEventListener('beforeunload', beforeUnload);
    return () => window.removeEventListener('beforeunload', beforeUnload);
  }, [shouldConfirm]);

  // Handle back/forward browser buttons
  useEffect(() => {
    if (!shouldConfirm) return;

    const handleBackButton = () => {
      if (!window.confirm(MESSAGE)) history.pushState(null, '', window.location.href);
      else {
        window.removeEventListener('popstate', handleBackButton);
        history.back();
        // When user clicks back after refreshing the page with unsaved changes
        setTimeout(() => {
          if (window.location.pathname === pathname) router.back();
        }, 0);
      }
    };

    // Push initial state to history to detect back navigation
    if (isInitialMount.current) {
      history.pushState(null, '', window.location.href);
      isInitialMount.current = false;
    }

    window.addEventListener('popstate', handleBackButton);
    return () => window.removeEventListener('popstate', handleBackButton);
  }, [shouldConfirm, pathname]);

  // Handle Next.js Link navigation
  useEffect(() => {
    if (!shouldConfirm) return;

    const handleClick = (e: MouseEvent) => {
      const target = e.target as HTMLElement;
      const anchor = target.closest('a');

      if (anchor && anchor.href && anchor.href.startsWith(window.location.origin)) {
        const targetUrl = new URL(anchor.href);
        const currentPath = pathname;

        if (targetUrl.pathname !== currentPath) {
          if (!window.confirm(MESSAGE)) {
            e.preventDefault();
            e.stopPropagation();
          }
        }
      }
    };

    document.addEventListener('click', handleClick, true);
    return () => document.removeEventListener('click', handleClick, true);
  }, [shouldConfirm, pathname]);
}
