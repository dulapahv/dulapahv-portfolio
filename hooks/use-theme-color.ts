"use client";

import { useEffect } from 'react';

import { useTheme } from 'next-themes';

export function useThemeColor() {
  const { resolvedTheme } = useTheme();

  useEffect(() => {
    // Get the theme color meta tag
    const themeColorMeta = document.querySelector('meta[name="theme-color"]');

    if (themeColorMeta) {
      // Update the color based on theme
      themeColorMeta.setAttribute(
        'content',
        resolvedTheme === 'dark' ? '#0a0a0a' : '#eef1f7',
      );
    }
  }, [resolvedTheme]);
}
