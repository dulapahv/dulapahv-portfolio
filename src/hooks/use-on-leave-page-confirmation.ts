"use client";

import { useEffect } from "react";

export function useOnLeavePageConfirmation(isDirty: boolean) {
  useEffect(() => {
    function beforeUnload(e: BeforeUnloadEvent) {
      if (!isDirty) return;
      e.preventDefault();
    }

    window.addEventListener("beforeunload", beforeUnload);

    return () => {
      window.removeEventListener("beforeunload", beforeUnload);
    };
  }, [isDirty]);
}
