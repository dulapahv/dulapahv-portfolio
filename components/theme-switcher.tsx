'use client';

import { useEffect, useRef, useState } from 'react';

import { Monitor, Moon, Sun } from 'lucide-react';
import { motion } from 'motion/react';
import { useTheme } from 'next-themes';

import { cn } from '@/lib/utils';
import { useMediaQuery } from '@/hooks/use-media-query';
import { useThemeColor } from '@/hooks/use-theme-color';

export const ThemeSwitcher = () => {
  const { theme, setTheme } = useTheme();
  const [mounted, setMounted] = useState(false);
  const [isExpanded, setIsExpanded] = useState(false);
  const [isHovered, setIsHovered] = useState(false);
  const [isFocused, setIsFocused] = useState(false);
  const containerRef = useRef<HTMLDivElement>(null);
  const isSmallScreen = useMediaQuery('(max-width: 678px)');
  const isTouchDevice = useMediaQuery('(hover: none) and (pointer: coarse)');

  // Expand when hovered, focused, or manually expanded on touch devices
  const expanded = isTouchDevice ? isExpanded : isHovered || isFocused;

  useThemeColor();

  useEffect(() => setMounted(true), []);

  // Close on click outside
  useEffect(() => {
    const handleClickOutside = (event: MouseEvent | TouchEvent) => {
      if (
        containerRef.current &&
        !containerRef.current.contains(event.target as Node)
      ) {
        setIsExpanded(false);
      }
    };

    if (isExpanded) {
      document.addEventListener('mousedown', handleClickOutside);
      document.addEventListener('touchstart', handleClickOutside);
      return () => {
        document.removeEventListener('mousedown', handleClickOutside);
        document.removeEventListener('touchstart', handleClickOutside);
      };
    }
  }, [isExpanded]);

  // Handle keyboard navigation
  const handleKeyDown = (event: React.KeyboardEvent) => {
    if (!containerRef.current) return;

    const buttons = containerRef.current.querySelectorAll('button');
    const currentIndex = Array.from(buttons).findIndex(
      (button) => button === document.activeElement,
    );

    switch (event.key) {
      case 'ArrowUp':
      case 'ArrowLeft':
        event.preventDefault();
        if (expanded && buttons.length > 1) {
          const newIndex =
            currentIndex > 0 ? currentIndex - 1 : buttons.length - 1;
          buttons[newIndex]?.focus();
        }
        break;
      case 'ArrowDown':
      case 'ArrowRight':
        event.preventDefault();
        if (expanded && buttons.length > 1) {
          const newIndex =
            currentIndex < buttons.length - 1 ? currentIndex + 1 : 0;
          buttons[newIndex]?.focus();
        }
        break;
      case 'Escape':
        event.preventDefault();
        if (isTouchDevice && isExpanded) {
          setIsExpanded(false);
          // Focus the main button after collapsing
          buttons[0]?.focus();
        }
        break;
      case 'Enter':
      case ' ':
        if (isTouchDevice && !expanded) {
          event.preventDefault();
          setIsExpanded(true);
        }
        break;
    }
  };

  // Track focus within the component
  const handleFocusIn = () => {
    setIsFocused(true);
  };

  const handleFocusOut = (event: React.FocusEvent) => {
    // Check if focus is moving outside the component
    if (!containerRef.current?.contains(event.relatedTarget as Node)) {
      setIsFocused(false);
      if (isTouchDevice) {
        setIsExpanded(false);
      }
    }
  };

  if (!mounted) return null;

  const themes = [
    {
      key: 'system',
      icon: Monitor,
      label: 'System theme',
    },
    {
      key: 'light',
      icon: Sun,
      label: 'Light theme',
    },
    {
      key: 'dark',
      icon: Moon,
      label: 'Dark theme',
    },
  ];

  const containerVariants = {
    hidden: {
      y: 100,
      opacity: 0,
    },
    visible: {
      y: 0,
      opacity: 1,
      transition: {
        type: 'spring',
        bounce: 0,
        duration: 0.5,
        delay: isHovered || isFocused || isExpanded ? 0 : 0.5,
      },
    },
  };

  const handleThemeChange = (newTheme: string) => {
    setTheme(newTheme);

    // Always collapse after theme selection for better UX
    if (isTouchDevice) {
      setIsExpanded(false);
    } else {
      // For non-touch devices, remove focus and reset states to collapse
      setIsFocused(false);
      setIsHovered(false);
      // Remove focus from the button to fully collapse
      if (document.activeElement instanceof HTMLElement) {
        document.activeElement.blur();
      }
    }
  };

  const handleContainerClick = () => {
    if (isTouchDevice) {
      setIsExpanded(!isExpanded);
    }
  };

  const currentTheme = themes.find((t) => t.key === theme);
  const CurrentIcon = currentTheme?.icon || Monitor;

  return (
    <motion.div
      ref={containerRef}
      className={cn(
        `ring-border bg-background-elevated/90 fixed right-5 bottom-4 z-50
        overflow-hidden rounded-full p-0.5 ring-1 backdrop-blur-xl`,
        isSmallScreen
          ? 'flex max-h-7 w-7 flex-col'
          : 'flex h-7 max-w-7 flex-row',
        theme === 'system' && 'justify-start',
        theme === 'light' && 'justify-center',
        theme === 'dark' && 'justify-end',
      )}
      variants={containerVariants}
      initial="hidden"
      animate={{
        ...containerVariants.visible,
        ...(isSmallScreen
          ? { maxHeight: expanded ? '100px' : '28px' }
          : { maxWidth: expanded ? '100px' : '28px' }),
      }}
      transition={{
        ...containerVariants.visible.transition,
        ...(isSmallScreen
          ? { maxHeight: { duration: 0.15, ease: 'easeOut' } }
          : { maxWidth: { duration: 0.15, ease: 'easeOut' } }),
      }}
      onMouseEnter={() => !isTouchDevice && setIsHovered(true)}
      onMouseLeave={() => !isTouchDevice && setIsHovered(false)}
      onClick={handleContainerClick}
      onFocusCapture={handleFocusIn}
      onBlurCapture={handleFocusOut}
      onKeyDown={handleKeyDown}
      role="group"
      aria-label="Theme selection"
      aria-expanded={expanded}
    >
      {/* For touch devices, show current theme icon when collapsed */}
      {isTouchDevice && !expanded ? (
        <motion.button
          type="button"
          className="hover:bg-background-subtle relative size-6 shrink-0 cursor-pointer rounded-full
            transition-colors"
          aria-label={`Current theme: ${currentTheme?.label}. Press Enter or Space to expand theme options`}
          aria-describedby="theme-instructions"
          whileTap={{ scale: 0.95 }}
        >
          <CurrentIcon className="text-foreground relative m-auto size-4" />
        </motion.button>
      ) : (
        // Show all theme options when expanded or on hover/focus
        themes.map(({ key, icon: Icon, label }, index) => {
          const isActive = theme === key;
          return (
            <motion.button
              type="button"
              key={key}
              className="hover:bg-background-subtle relative size-6 shrink-0 cursor-pointer rounded-full
                transition-colors"
              onClick={(e) => {
                e.stopPropagation();
                handleThemeChange(key);
              }}
              aria-label={`${label}${isActive ? ', currently selected' : ''}`}
              aria-pressed={isActive}
              aria-describedby="theme-instructions"
              whileHover={{ scale: 1.1 }}
              whileTap={{ scale: 0.95 }}
            >
              {isActive && (
                <motion.div
                  layoutId="activeTheme"
                  className="bg-background-muted border-border-subtle absolute inset-0 rounded-full border
                    shadow-sm"
                  transition={{ type: 'spring', duration: 0.5 }}
                />
              )}
              <Icon
                className={cn(
                  'relative m-auto size-4',
                  isActive
                    ? 'text-foreground'
                    : 'text-foreground-subtle hover:text-foreground-muted transition-colors',
                )}
                aria-hidden="true"
              />
            </motion.button>
          );
        })
      )}

      {/* Screen reader instructions */}
      <div id="theme-instructions" className="sr-only">
        {isTouchDevice
          ? 'Use Enter or Space to expand, arrow keys to navigate between themes, Escape to close'
          : 'Use arrow keys to navigate between themes'}
      </div>

      {/* Live region for theme changes */}
      <div aria-live="polite" aria-atomic="true" className="sr-only">
        {mounted && `Theme changed to ${currentTheme?.label}`}
      </div>
    </motion.div>
  );
};
