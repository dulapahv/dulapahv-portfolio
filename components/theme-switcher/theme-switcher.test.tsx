import { act, render, screen, waitFor } from '@testing-library/react';
import { useTheme } from 'next-themes';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { useMediaQuery } from '@/hooks/use-media-query';

import { ThemeSwitcher } from './theme-switcher';

// Mock dependencies
vi.mock('next-themes', () => ({
  useTheme: vi.fn()
}));

vi.mock('@/hooks/use-media-query', () => ({
  useMediaQuery: vi.fn()
}));

// Mock framer-motion
vi.mock('motion/react', () => ({
  motion: {
    /* eslint-disable  @typescript-eslint/no-explicit-any */
    div: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    /* eslint-disable  @typescript-eslint/no-explicit-any */
    button: ({ children, ...props }: any) => <button {...props}>{children}</button>
  }
}));

describe('ThemeSwitcher', () => {
  const mockSetTheme = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(useTheme).mockReturnValue({
      theme: 'system',
      setTheme: mockSetTheme,
      themes: ['light', 'dark', 'system'],
      systemTheme: 'light',
      resolvedTheme: 'light',
      forcedTheme: undefined
    });
    vi.mocked(useMediaQuery).mockReturnValue(false);
  });

  describe('Rendering', () => {
    it('should render empty before mounted', () => {
      // The component returns null before mounting
      // In tests, it mounts synchronously so we'll see content
      const { container } = render(<ThemeSwitcher />);
      expect(container).toBeInTheDocument();
    });

    it('should render after mounting', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const group = screen.getByRole('group', { name: /theme selection/i });
        expect(group).toBeInTheDocument();
      });
    });

    it('should render all theme buttons when not on touch device', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        expect(screen.getByRole('button', { name: /system theme/i })).toBeInTheDocument();
        expect(screen.getByRole('button', { name: /light theme/i })).toBeInTheDocument();
        expect(screen.getByRole('button', { name: /dark theme/i })).toBeInTheDocument();
      });
    });

    it('should render current theme button on touch device when collapsed', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(hover: none) and (pointer: coarse)') return true;
        return false;
      });

      render(<ThemeSwitcher />);

      await waitFor(() => {
        const button = screen.getByRole('button');
        expect(button).toHaveAttribute('aria-label', expect.stringContaining('Current theme'));
      });
    });
  });

  describe('Theme switching', () => {
    it('should call setTheme when clicking system theme button', async () => {
      vi.mocked(useTheme).mockReturnValue({
        theme: 'light',
        setTheme: mockSetTheme,
        themes: ['light', 'dark', 'system'],
        systemTheme: 'light',
        resolvedTheme: 'light',
        forcedTheme: undefined
      });

      render(<ThemeSwitcher />);

      await waitFor(() => {
        const systemButton = screen.getByRole('button', { name: /system theme/i });
        systemButton.click();
      });

      expect(mockSetTheme).toHaveBeenCalledWith('system');
    });

    it('should call setTheme when clicking light theme button', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const lightButton = screen.getByRole('button', { name: /light theme/i });
        lightButton.click();
      });

      expect(mockSetTheme).toHaveBeenCalledWith('light');
    });

    it('should call setTheme when clicking dark theme button', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const darkButton = screen.getByRole('button', { name: /dark theme/i });
        darkButton.click();
      });

      expect(mockSetTheme).toHaveBeenCalledWith('dark');
    });
  });

  describe('Active state', () => {
    it('should mark system theme as active', async () => {
      vi.mocked(useTheme).mockReturnValue({
        theme: 'system',
        setTheme: mockSetTheme,
        themes: ['light', 'dark', 'system'],
        systemTheme: 'light',
        resolvedTheme: 'light',
        forcedTheme: undefined
      });

      render(<ThemeSwitcher />);

      await waitFor(() => {
        const systemButton = screen.getByRole('button', {
          name: /system theme.*currently selected/i
        });
        expect(systemButton).toHaveAttribute('aria-pressed', 'true');
      });
    });

    it('should mark light theme as active', async () => {
      vi.mocked(useTheme).mockReturnValue({
        theme: 'light',
        setTheme: mockSetTheme,
        themes: ['light', 'dark', 'system'],
        systemTheme: 'light',
        resolvedTheme: 'light',
        forcedTheme: undefined
      });

      render(<ThemeSwitcher />);

      await waitFor(() => {
        const lightButton = screen.getByRole('button', {
          name: /light theme.*currently selected/i
        });
        expect(lightButton).toHaveAttribute('aria-pressed', 'true');
      });
    });

    it('should mark dark theme as active', async () => {
      vi.mocked(useTheme).mockReturnValue({
        theme: 'dark',
        setTheme: mockSetTheme,
        themes: ['light', 'dark', 'system'],
        systemTheme: 'light',
        resolvedTheme: 'dark',
        forcedTheme: undefined
      });

      render(<ThemeSwitcher />);

      await waitFor(() => {
        const darkButton = screen.getByRole('button', { name: /dark theme.*currently selected/i });
        expect(darkButton).toHaveAttribute('aria-pressed', 'true');
      });
    });
  });

  describe('Touch device behavior', () => {
    beforeEach(() => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(hover: none) and (pointer: coarse)') return true;
        return false;
      });
    });

    it('should expand when clicking container on touch device', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const container = screen.getByRole('group');
        expect(container).toHaveAttribute('aria-expanded', 'false');
      });

      const container = screen.getByRole('group');
      container.click();

      await waitFor(() => {
        expect(container).toHaveAttribute('aria-expanded', 'true');
      });
    });

    it('should collapse after selecting theme on touch device', async () => {
      render(<ThemeSwitcher />);

      // Expand first
      await waitFor(() => {
        const container = screen.getByRole('group');
        container.click();
      });

      await waitFor(() => {
        const container = screen.getByRole('group');
        expect(container).toHaveAttribute('aria-expanded', 'true');
      });

      // Click theme button
      const lightButton = screen.getByRole('button', { name: /light theme/i });
      lightButton.click();

      await waitFor(() => {
        const container = screen.getByRole('group');
        expect(container).toHaveAttribute('aria-expanded', 'false');
      });
    });
  });

  describe('Keyboard navigation', () => {
    it('should navigate with ArrowRight key', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const buttons = screen.getAllByRole('button');
        buttons[0].focus();
      });

      const container = screen.getByRole('group');
      container.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowRight', bubbles: true }));

      await waitFor(() => {
        const buttons = screen.getAllByRole('button');
        expect(document.activeElement).toBe(buttons[1]);
      });
    });

    it('should navigate with ArrowLeft key', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const buttons = screen.getAllByRole('button');
        buttons[1].focus();
      });

      const container = screen.getByRole('group');
      container.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowLeft', bubbles: true }));

      await waitFor(() => {
        const buttons = screen.getAllByRole('button');
        expect(document.activeElement).toBe(buttons[0]);
      });
    });

    it('should wrap around when navigating with ArrowRight from last button', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const buttons = screen.getAllByRole('button');
        buttons[buttons.length - 1].focus();
      });

      const container = screen.getByRole('group');
      container.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowRight', bubbles: true }));

      await waitFor(() => {
        const buttons = screen.getAllByRole('button');
        expect(document.activeElement).toBe(buttons[0]);
      });
    });

    it('should handle Enter key on touch device to expand', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(hover: none) and (pointer: coarse)') return true;
        return false;
      });

      render(<ThemeSwitcher />);

      await waitFor(() => {
        const container = screen.getByRole('group');
        expect(container).toHaveAttribute('aria-expanded', 'false');
      });

      const container = screen.getByRole('group');
      container.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true }));

      await waitFor(() => {
        expect(container).toHaveAttribute('aria-expanded', 'true');
      });
    });

    it('should handle Space key on touch device to expand', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(hover: none) and (pointer: coarse)') return true;
        return false;
      });

      render(<ThemeSwitcher />);

      await waitFor(() => {
        const container = screen.getByRole('group');
        expect(container).toHaveAttribute('aria-expanded', 'false');
      });

      const container = screen.getByRole('group');
      container.dispatchEvent(new KeyboardEvent('keydown', { key: ' ', bubbles: true }));

      await waitFor(() => {
        expect(container).toHaveAttribute('aria-expanded', 'true');
      });
    });

    it('should handle Escape key on touch device to collapse', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(hover: none) and (pointer: coarse)') return true;
        return false;
      });

      render(<ThemeSwitcher />);

      // First expand
      await waitFor(() => {
        const container = screen.getByRole('group');
        container.click();
      });

      await waitFor(() => {
        const container = screen.getByRole('group');
        expect(container).toHaveAttribute('aria-expanded', 'true');
      });

      // Then press Escape
      const container = screen.getByRole('group');
      container.dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape', bubbles: true }));

      await waitFor(() => {
        expect(container).toHaveAttribute('aria-expanded', 'false');
      });
    });
  });

  describe('Click outside behavior', () => {
    it('should close when clicking outside on touch device', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(hover: none) and (pointer: coarse)') return true;
        return false;
      });

      render(<ThemeSwitcher />);

      // Expand first
      await waitFor(() => {
        const container = screen.getByRole('group');
        container.click();
      });

      await waitFor(() => {
        const container = screen.getByRole('group');
        expect(container).toHaveAttribute('aria-expanded', 'true');
      });

      // Click outside
      act(() => {
        document.dispatchEvent(new MouseEvent('mousedown', { bubbles: true }));
      });

      await waitFor(() => {
        const container = screen.getByRole('group');
        expect(container).toHaveAttribute('aria-expanded', 'false');
      });
    });
  });

  describe('Accessibility', () => {
    it('should have aria-label for theme selection', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const group = screen.getByRole('group', { name: /theme selection/i });
        expect(group).toHaveAttribute('aria-label', 'Theme selection');
      });
    });

    it('should have aria-expanded attribute', async () => {
      render(<ThemeSwitcher />);

      await waitFor(() => {
        const group = screen.getByRole('group');
        expect(group).toHaveAttribute('aria-expanded');
      });
    });

    it('should have screen reader instructions for touch devices', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(hover: none) and (pointer: coarse)') return true;
        return false;
      });

      const { container } = render(<ThemeSwitcher />);

      await waitFor(() => {
        const instructions = container.querySelector('#theme-instructions');
        expect(instructions?.textContent).toContain('Enter or Space to expand');
      });
    });

    it('should have screen reader instructions for non-touch devices', async () => {
      const { container } = render(<ThemeSwitcher />);

      await waitFor(() => {
        const instructions = container.querySelector('#theme-instructions');
        expect(instructions?.textContent).toContain('arrow keys to navigate');
      });
    });

    it('should have live region for theme changes', async () => {
      const { container } = render(<ThemeSwitcher />);

      await waitFor(() => {
        const liveRegion = container.querySelector('[aria-live="polite"]');
        expect(liveRegion).toBeInTheDocument();
      });
    });
  });

  describe('Responsive behavior', () => {
    it('should apply small screen styles', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(max-width: 678px)') return true;
        return false;
      });

      const { container } = render(<ThemeSwitcher />);

      await waitFor(() => {
        const group = container.querySelector('[role="group"]');
        expect(group?.className).toContain('flex-col');
      });
    });

    it('should apply large screen styles', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(max-width: 678px)') return false;
        return false;
      });

      const { container } = render(<ThemeSwitcher />);

      await waitFor(() => {
        const group = container.querySelector('[role="group"]');
        expect(group?.className).toContain('flex-row');
      });
    });
  });

  describe('Snapshot', () => {
    it('should match snapshot with system theme', async () => {
      vi.mocked(useTheme).mockReturnValue({
        theme: 'system',
        setTheme: mockSetTheme,
        themes: ['light', 'dark', 'system'],
        systemTheme: 'light',
        resolvedTheme: 'light',
        forcedTheme: undefined
      });

      const { container } = render(<ThemeSwitcher />);

      await waitFor(() => {
        expect(screen.getByRole('group')).toBeInTheDocument();
      });

      expect(container).toMatchSnapshot();
    });
  });
});
