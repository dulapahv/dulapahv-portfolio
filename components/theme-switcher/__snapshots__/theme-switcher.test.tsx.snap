// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeSwitcher > Snapshot > should match snapshot with system theme 1`] = `
<div>
  <div
    animate="[object Object]"
    aria-expanded="false"
    aria-label="Theme selection"
    class="ring-border bg-background-elevated/90 fixed right-5 bottom-4 z-50 overflow-hidden rounded-full p-0.5 ring-1 backdrop-blur-xl flex h-7 max-w-7 flex-row justify-start"
    initial="hidden"
    role="group"
    transition="[object Object]"
    variants="[object Object]"
  >
    <button
      aria-describedby="theme-instructions"
      aria-label="System theme, currently selected"
      aria-pressed="true"
      class="relative size-6 shrink-0 cursor-pointer rounded-full transition-colors hover:bg-background-subtle"
      type="button"
      whilehover="[object Object]"
      whiletap="[object Object]"
    >
      <div
        class="bg-background-muted border-border-subtle absolute inset-0 rounded-full border shadow-sm"
        layoutid="activeTheme"
        transition="[object Object]"
      />
      <svg
        aria-hidden="true"
        class="relative m-auto size-4 text-foreground-subtle hover:text-foreground-muted transition-colors"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M208,40H48A24,24,0,0,0,24,64V176a24,24,0,0,0,24,24H208a24,24,0,0,0,24-24V64A24,24,0,0,0,208,40Zm8,136a8,8,0,0,1-8,8H48a8,8,0,0,1-8-8V64a8,8,0,0,1,8-8H208a8,8,0,0,1,8,8Zm-48,48a8,8,0,0,1-8,8H96a8,8,0,0,1,0-16h64A8,8,0,0,1,168,224Z"
        />
      </svg>
    </button>
    <button
      aria-describedby="theme-instructions"
      aria-label="Light theme"
      aria-pressed="false"
      class="relative size-6 shrink-0 cursor-pointer rounded-full transition-colors hover:bg-background-subtle"
      type="button"
      whilehover="[object Object]"
      whiletap="[object Object]"
    >
      <svg
        aria-hidden="true"
        class="relative m-auto size-4 text-foreground-subtle hover:text-foreground-muted transition-colors"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M120,40V16a8,8,0,0,1,16,0V40a8,8,0,0,1-16,0Zm72,88a64,64,0,1,1-64-64A64.07,64.07,0,0,1,192,128Zm-16,0a48,48,0,1,0-48,48A48.05,48.05,0,0,0,176,128ZM58.34,69.66A8,8,0,0,0,69.66,58.34l-16-16A8,8,0,0,0,42.34,53.66Zm0,116.68-16,16a8,8,0,0,0,11.32,11.32l16-16a8,8,0,0,0-11.32-11.32ZM192,72a8,8,0,0,0,5.66-2.34l16-16a8,8,0,0,0-11.32-11.32l-16,16A8,8,0,0,0,192,72Zm5.66,114.34a8,8,0,0,0-11.32,11.32l16,16a8,8,0,0,0,11.32-11.32ZM48,128a8,8,0,0,0-8-8H16a8,8,0,0,0,0,16H40A8,8,0,0,0,48,128Zm80,80a8,8,0,0,0-8,8v24a8,8,0,0,0,16,0V216A8,8,0,0,0,128,208Zm112-88H216a8,8,0,0,0,0,16h24a8,8,0,0,0,0-16Z"
        />
      </svg>
    </button>
    <button
      aria-describedby="theme-instructions"
      aria-label="Dark theme"
      aria-pressed="false"
      class="relative size-6 shrink-0 cursor-pointer rounded-full transition-colors hover:bg-background-subtle"
      type="button"
      whilehover="[object Object]"
      whiletap="[object Object]"
    >
      <svg
        aria-hidden="true"
        class="relative m-auto size-4 text-foreground-subtle hover:text-foreground-muted transition-colors"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M233.54,142.23a8,8,0,0,0-8-2,88.08,88.08,0,0,1-109.8-109.8,8,8,0,0,0-10-10,104.84,104.84,0,0,0-52.91,37A104,104,0,0,0,136,224a103.09,103.09,0,0,0,62.52-20.88,104.84,104.84,0,0,0,37-52.91A8,8,0,0,0,233.54,142.23ZM188.9,190.34A88,88,0,0,1,65.66,67.11a89,89,0,0,1,31.4-26A106,106,0,0,0,96,56,104.11,104.11,0,0,0,200,160a106,106,0,0,0,14.92-1.06A89,89,0,0,1,188.9,190.34Z"
        />
      </svg>
    </button>
    <div
      class="sr-only"
      id="theme-instructions"
    >
      Use arrow keys to navigate between themes
    </div>
    <div
      aria-atomic="true"
      aria-live="polite"
      class="sr-only"
    >
      Theme changed to System theme
    </div>
  </div>
</div>
`;
